datafusion/expr-common/src/scalar.rs

```diff
@@ -116,6 +143,10 @@ impl Scalar {
         ScalarValue::iter_to_array_of_type(scalars.map(|scalar| scalar.value), &data_type)
     }
 
+    pub fn cast_to(&self, data_type: &DataType) -> Result<Self> {
```

Comment:
I think this function is useful to keep here because we can then enforce the same constraints introduced in the construction of the struct and potentially support logic like 

`data_type.is_logically_equivalent_to(self.data_type)` => `self.data_type = data_type` (skipping the cast)